quartz-core/src/main/java/org/quartz/CronExpression.java

```diff
@@ -1586,11 +1591,22 @@ protected void setCalendarHour(Calendar cal, int hour) {
     }
 
     /**
-     * NOT YET IMPLEMENTED: Returns the time before the given time
+     * Returns the time before the given time
      * that the <code>CronExpression</code> matches.
      */ 
     public Date getTimeBefore(Date endTime) { 
-        // FUTURE_TODO: implement QUARTZ-423
+        final Date fireTimeAfterOrigin = getTimeAfter(endTime);
+        if (fireTimeAfterOrigin == null) return null;
+        for (final long interval : GET_TIME_BEFORE_INTERVALS) {
```

Comment:
This looks like it wants to be two binary searches in opposite directions. The outer loop is essentially searching backwards in ever widening circles trying to establish a lower bound on the time we want. The inner loop is then searching forwards trying to find the time we want. It would seem that it would be more efficient and clearer if we made the first search use an actual exponential growth pattern, and then once we've established a lower and upper bound we can do a binary search from that point. The following does exactly that:
```java
    public Date getTimeBefore(Date endTime) {
        final Date after = getTimeAfter(endTime);

        if (after == null) return null;

        long afterMillis = after.getTime();

        long upper = afterMillis;
        long lower = Long.MIN_VALUE;

        //establish a lower bound (and refine the upper while we're at it)
        for (long i = 1L; i > 0; i <<= 1L) {
            long candidateMillis = afterMillis - i;
            Date candidate = getTimeAfter(new Date(candidateMillis));
            if (candidate == null) {
                return null;
            }
            if (candidate.equals(after)) {
                upper = candidateMillis - 1;
            } else {
                lower = candidateMillis;
                break;
            }
        }

        //do a binary search for the threshold value
        while (lower < (upper - 1)) {
            long middle = (lower + upper) >>> 1;

            Date candidate = getTimeAfter(new Date(middle));
            if (candidate == null) {
                return null;
            }

            if (candidate.equals(after)) {
                upper = middle - 1;
            } else {
                lower = middle;
            }
        }
        return getTimeAfter(new Date(lower));
    }
```

Reply:
The above is overly precise and therefore too costly... please ignore.